File: api/client/src/main/java/su/plo/voice/api/client/audio/device/OutputDevice.java
package su.plo.voice.api.client.audio.device;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import su.plo.voice.api.client.audio.device.source.DeviceSource;
import su.plo.voice.api.util.Params;

import javax.sound.sampled.AudioFormat;

// todo: doc
public interface OutputDevice<T extends DeviceSource> extends AudioDevice {

    /**
     * Create a new source
     *
     * @param stereo if true source will be stereo
     * @param params parameters
     *
     * @return the device source
     */
    T createSource(boolean stereo, @NotNull Params params) throws DeviceException;

    /**
     * Reloads the device
     *
     * @param format audio format
     *               if null current audio format will be used
     * @param params device params
     *               will use current audio params and overwrite it with yours
     *
     * @throws DeviceException if the device cannot be reloaded
     */
    void reload(@Nullable AudioFormat format, @NotNull Params params) throws DeviceException;

    @Override
    default DeviceType getType() {
        return DeviceType.OUTPUT;
    }
}

File: api/client/src/main/kotlin/su/plo/voice/api/client/audio/device/OutputDevice.kt
package su.plo.voice.api.client.audio.device

import su.plo.voice.api.client.audio.device.source.DeviceSource
import su.plo.voice.api.client.audio.device.source.DeviceSourceParams
import java.util.concurrent.CompletableFuture

/**
 * An audio output device capable of creating and managing audio sources.
 *
 * @param S The type of audio sources that can be created by this output device.
 */
interface OutputDevice<S : DeviceSource> : AudioDevice {

    /**
     * Creates a new audio source associated with this output device.
     *
     * @param stereo If `true`, the created source will be stereo, otherwise - mono.
     * @param params Parameters specific to the device source configuration.
     *
     * @return The device source that has been created.
     *
     * @throws DeviceException If there is an issue creating the audio source.
     */
    @Throws(DeviceException::class)
    fun createSource(stereo: Boolean, params: DeviceSourceParams): S

    /**
     * Closes all audio sources associated with this output device.
     */
    suspend fun closeSources()

    /**
     * Asynchronously closes all audio sources associated with this output device.
     *
     * @return A [CompletableFuture] that represents the completion of the operation.
     */
    fun closeSourcesAsync(): CompletableFuture<Void?>

    override val type: DeviceType
        get() = DeviceType.OUTPUT
}
